dipy/align/mattes.pyx

```diff
@@ -0,0 +1,1390 @@
+            if (self.metric_grad is None) or (self.metric_grad.shape[0] != sh):
+                self.metric_grad = np.empty(sh)
+            grad = self.metric_grad
+
+        self.metric_val = _compute_mattes_mi(self.joint, self.joint_grad,
+                                             self.smarginal, self.mmarginal,
+                                             grad)
+
+
+cdef inline double _bin_normalize(double x, double mval, double delta) nogil:
+    r''' Normalizes intensity x to the range covered by the Parzen histogram
+    We assume that mval was computed as:
+
+    (1) mval = xmin / delta - padding
+
+    where xmin is the minimum observed image intensity and delta is the
+    bin size, computed as:
+
+    (2) delta = (xmax - xmin)/(nbins - 2 * padding)
+
+    If the minimum and maximum intensities were assigned to the first and last
+    bins (with no padding), it could be possible that samples at the first and
+    last bins contribute to "non-existing" bins beyond the boundary (because
+    the support of the Parzen window may be larger than one bin). The padding
+    bins are used to collect such contributions (i.e. the probability of
+    observing a value beyond the minimum and maximum observed intensities may
+    correctly be assigned a positive value).
+
+    The normalized intensity is (from eq(1) ):
+
+    (3) nx = (x - xmin) / delta + padding = x / delta - mval
+
+    This means that normalized intensity nx must lie in the closed interval
+    [padding, nbins-padding], which contains bins with indices
+    padding, padding+1, ..., nbins - 1 - padding (i.e., nbins - 2*padding bins)
+
+    '''
+    return x / delta - mval
+
+
+cdef inline cnp.npy_intp _bin_index(double normalized, int nbins,
+                                    int padding) nogil:
+    r''' Index of the bin in which the normalized intensity 'normalized' lies
+    The intensity is assumed to have been normalized to the range of intensities
+    covered by the histogram: the bin index is the integer part of the argument,
+    which must be within the interval [padding, nbins - 1 - padding].
+
+    Parameters
+    ----------
+    normalized : float
+        normalized intensity
+    nbins : int
+        number of histogram bins
+    padding : int
+        number of bins used as padding (the total bins used for padding at
+        both sides of the histogram is actually 2*padding)
+
+    Returns
+    -------
+    bin : int
+        index of the bin in which the normalized intensity 'normalized' lies
+    '''
+    cdef:
+        cnp.npy_intp bin
+
+    bin = <cnp.npy_intp>(normalized)
+    if bin < padding:
+        return padding
+    if bin > nbins - 1 - padding:
+        return nbins - 1 - padding
+    return bin
+
+
+def cubic_spline(double[:] x):
+    r''' Evaluates the cubic spline at a set of values
+
+    Parameters
+    ----------
+    x : array, shape (n)
+        input values
+    '''
+    cdef:
+        int i
+        int n = x.shape[0]
```

Comment:
Yeah... this habit of declaring "everything" as int is going to be hard to avoid =) (now I will declare everything as cnp.npy_intp, just in case hehe ;) )